src/_articles/libraries/dart-io.md

```diff
@@ -0,0 +1,386 @@
+by going through a couple of examples.
+
+Dart is a single-threaded programming language.
+If an operation blocks the Dart thread,
+the application makes no progress before that operation completes.
+For scalability it is therefore crucial that no I/O operations block.
+Instead of blocking on I/O operations,
+dart:io uses an asynchronous programming model inspired by
+[node.js,](http://nodejs.org)
+[EventMachine,](https://github.com/eventmachine/eventmachine/wiki) and
+[Twisted.](http://twistedmatrix.com/trac/)
+
+## The Dart VM and the event loop
+
+Before we dive into asynchronous I/O operations,
+it might be useful to explain how the standalone Dart VM operates.
+
+When executing a server-side application,
+the Dart VM runs in an event loop with
+an associated event queue of pending asynchronous operations.
+The VM terminates when it has executed the current code to completion
+and no more pending operations are in the queue.
+
+One simple way to add an event to the event queue is to
+schedule a function to be called in the future.
+You can do this by creating a
+[Timer]({{site.dart_api}}/{{site.data.pkg-vers.SDK.channel}}/dart-async/Timer-class.html) object.
+The following example registers a timer with the event queue
+and then drops off the end of main().
+Because a pending operation is in the event queue,
+the VM does not terminate at that point.
+After one second,
+the timer fires and the code in the argument callback executes.
+Once that code executes to completion,
+no more pending operations are in the event queue
+and the VM terminates.
+
+<?code-excerpt "misc/lib/articles/io/io_timer_test.dart"?>
+{% prettify dart %}
+import 'dart:async';
+
+void main() {
+  Timer(Duration(seconds: 1), () => print('timer'));
+  print('end of main');
+}
+{% endprettify %}
+
+Running this example at the command line, we get:
+
+```terminal
+$ dart timer.dart
+end of main
+timer
+```
+
+Had we made the timer repeating by using the Timer.periodic() constructor,
+the VM would not terminate
+and would continue to print out 'timer' every second.
+
+## File system access
+
+The dart:io library provides access to files and directories through the
+[File]({{site.dart_api}}/{{site.data.pkg-vers.SDK.channel}}/dart-io/File-class.html) and
+[Directory]({{site.dart_api}}/{{site.data.pkg-vers.SDK.channel}}/dart-io/Directory-class.html) classes.
+
+The following example prints its own source code.
+To determine the location of the source code being executed,
+we use the
+[Platform]({{site.dart_api}}/{{site.data.pkg-vers.SDK.channel}}/dart-io/Platform-class.html)
+class.
+
+
+<?code-excerpt "misc/lib/articles/io/io_file_system_test.dart"?>
+{% prettify dart %}
+import 'dart:async' show Future;
+import 'dart:convert';
+import 'dart:io';
+
+Future<void> main() async {
+  var file = File(Platform.script.toFilePath());
+  print("${await (file.readAsString(encoding: ascii))}");
+}
+{% endprettify %}
+
+Notice that the readAsString() method is asynchronous;
```

Comment:
Maybe quote `readAsString()` ?

Reply:
We used to have a policy on dartlang.org of avoiding code font when possible, but that contradicts the Google developer docs style guide (https://developers.google.com/style/code-in-text) and what many developers expect. So we're updating as we go. I'll add code font to this article.